src/hooks/useSigningKeys.ts

```diff
@@ -0,0 +1,86 @@
+import { ethers } from "ethers";
+import { useState, useEffect, useContext } from "react";
+import { ErrorContext } from "common/context/ErrorContext";
+
+import { recoverPublicKey } from "features/vote/helpers/recoverPublicKey";
+import { derivePrivateKey } from "features/vote/helpers/derivePrivateKey";
+import usePrevious from "common/hooks/usePrevious";
+
+// User has to sign once per round.
+// public and private signing keys is indexed by address, and we ask them to resign if the roundMessage changes.
+export interface SigningKeys {
+  [address: string]: {
+    publicKey: string;
+    privateKey: string;
+    roundMessage: string;
+  };
+}
+
+export default function useSigningKeys(
+  signer: ethers.Signer | null,
+  address: string | null,
+  roundId: string
+) {
+  const [signingKeys, setSigningKeys] = useState<SigningKeys>(() => {
+    const storageKeys = localStorage.getItem("signingKeys");
+    return storageKeys ? (JSON.parse(storageKeys) as SigningKeys) : {};
+  });
+
+  const { addError } = useContext(ErrorContext);
+
+  const previousSigner = usePrevious(signer);
+  const previousAddress = usePrevious(address);
+
+  useEffect(() => {
+    if (
+      signer &&
+      address &&
+      roundId &&
+      (previousAddress === null || previousSigner === null)
+    ) {
+      const message = `UMA Protocol one time key for round: ${roundId}`;
+      const keyExists = signingKeys[address];
+      if (!keyExists || keyExists.roundMessage !== message) {
+        signer
+          .signMessage(message)
+          .then((msg) => {
+            const key = {} as {
+              publicKey: string;
+              privateKey: string;
+              roundMessage: string;
+            };
+
+            const privateKey = derivePrivateKey(msg);
+            const publicKey = recoverPublicKey(privateKey);
+            key.privateKey = privateKey;
+            key.publicKey = publicKey;
+            key.roundMessage = message;
+
+            setSigningKeys((prevKeys) => {
+              const updateKeys = { ...prevKeys, [address]: key };
+              localStorage.setItem("signingKeys", JSON.stringify(updateKeys));
+
+              return updateKeys;
+            });
+          })
+          .catch((err) => {
+            const error = new Error("Sign failed.");
+            addError(error);
+          });
+      }
+    }
+  }, [
+    signer,
+    address,
+    signingKeys,
+    addError,
```

Comment:
Is this a function? Are we sure this doesn't get compared by reference and causes the useEffect to run on every render because the function gets reinitialized on each run?

Reply:
It shouldn't, as I believe addError is memoized.

Reply:
Would be nice to verify. It would be great, in general, like for us to have better reasoning around how often these useEffects run (and add comments about it!) and what types of changes trigger them. Otherwise, it's difficult to reason about the state changes that can cause the ifs to block or accept changes.